### PR TITLE
Fix kernels/nms: race on keep[], uninitialized read, wrong index semantics

### LINKED ISSUE #517: [Bug] kernels/nms: CUDA kernel races on keep[] and returns wrong indices — results nondeterministic and inconsistent with torchvision.ops.nms
﻿## Summary

`nms_kernel` in `kernels/nms/nms.cu` has three compounding correctness defects:

1. **Read of uninitialized memory**: the `keep` tensor is allocated with `torch::empty`, but the kernel reads other threads' `keep[i]` entries as the suppression criterion — reading garbage;
2. **Cross-block data race**: each thread sequentially scans `keep[0..idx-1]`, while those entries are concurrently written by threads in other blocks with no synchronization whatsoever — the same input produces different results across runs;
3. **Wrong index semantics**: the returned indices refer to the score-sorted array, but are never mapped back to the original input indices — inconsistent with `torchvision.ops.nms`.

Reproduced: 5 runs on the same fixed input produce 2 different results; none of them is correct (see minimal reproduction below).

## Environment

| Item | Value |
|---|---|
| GPU | NVIDIA GeForce RTX 4090 (24 GB, sm_89) |
| OS | Ubuntu (x86_64) |
| CUDA Toolkit | nvcc 12.4 |
| PyTorch | 2.5.1+cu124 |
| torchvision | 0.20.1+cu124 (used as the ground-truth `torchvision.ops.nms`) |

## Minimal Reproduction

Fixed 6-box test case (strictly decreasing scores, no ties). Coordinates and hand-computed IoU relations:

| idx | box (x1,y1,x2,y2) | score | IoU vs | IoU value | expected |
|---|---|---|---|---|---|
| 0 | (0, 0, 10, 10) | 0.9 | — | — | ✅ keep |
| 1 | (1, 1, 11, 11) | 0.8 | box0 | 81/119 ≈ **0.68** | ❌ suppress (>0.5) |
| 2 | (20, 20, 30, 30) | 0.7 | — | — | ✅ keep |
| 3 | (2, 2, 12, 12) | 0.6 | box0 | 64/136 ≈ **0.47** | ✅ keep (≤0.5) |
| 4 | (40, 40, 50, 50) | 0.5 | — | — | ✅ keep |
| 5 | (21, 21, 31, 31) | 0.4 | box2 | 81/119 ≈ **0.68** | ❌ suppress (>0.5) |

**Correct output (`torchvision.ops.nms`)**: `keep = [0, 2, 3, 4]`

Reproduction script (run in `kernels/nms/`):

```python
import torch
from torch.utils.cpp_extension import load
from torchvision.ops import nms as tv_nms

lib = load(name="nms_lib", sources=["nms.cu"],
           extra_cuda_cflags=["-O3", "--use_fast_math"],
           extra_cflags=["-std=c++17"])

boxes = torch.tensor([
    [0.0,  0.0,  10.0, 10.0],   # 0
    [1.0,  1.0,  11.0, 11.0],   # 1  IoU(0,1)=0.68 -> suppress
    [20.0, 20.0, 30.0, 30.0],   # 2
    [2.0,  2.0,  12.0, 12.0],   # 3  IoU(0,3)=0.47 -> keep
    [40.0, 40.0, 50.0, 50.0],   # 4
    [21.0, 21.0, 31.0, 31.0],   # 5  IoU(2,5)=0.68 -> suppress
], dtype=torch.float32).cuda()
scores = torch.tensor([0.9, 0.8, 0.7, 0.6, 0.5, 0.4],
                      dtype=torch.float32).cuda()

print("torchvision :", tv_nms(boxes, scores, 0.5).cpu().tolist())
for run in range(5):
    print(f"lib.nms #{run} :", lib.nms(boxes, scores, 0.5).cpu().tolist())
```

**Actual output**:

```
torchvision : [0, 2, 3, 4]
lib.nms #0  : [0, 1, 2, 3, 4]
lib.nms #1  : [0, 2, 3, 4]
lib.nms #2  : [0, 2, 3, 4]
lib.nms #3  : [0, 2, 3, 4]
lib.nms #4  : [0, 2, 3, 4]
```

- **Nondeterminism**: 5 consecutive calls in the same process on the same input — run #0 differs from the other 4 → direct evidence of the race;
- **Incorrectness**: in `#1..#4`, box1 (IoU=0.68 > 0.5) is not suppressed; in `#0`, box5 survives as well. None of the 5 outputs is fully correct.

### 1. `keep` is read before being initialized

Host side:

```cpp
auto keep = torch::empty({boxes.size(0)}, toption);   // <-- empty = garbage memory
```

Device side — thread `idx` reads `keep[i]` (`i < idx`) to decide whether box i is already suppressed:

```cpp
for (int i = 0; i < idx; ++i) {
    if (keep[i] == 0) continue;   // <-- garbage value, or a value thread i has not written yet
    ...
}
```

`torch::empty` performs no zero-initialization. When thread `idx` reads `keep[i]`, thread `i` may not have reached its write yet, so the read returns whatever garbage the allocator left there → the suppression check misfires (this is why box1 survives in `#1..#4`).

### 2. Cross-block data race

The kernel launches as `block=32, grid=ceil(N/32)`, so the writer of `keep[i]` and the reader in thread `idx` usually live in **different blocks**. CUDA provides no implicit synchronization across blocks — even if the host used `torch::zeros` to eliminate the garbage values, "has thread i written keep[i] yet?" remains an unsynchronized concurrent read/write (data race, UB) for thread `idx` whenever `i < idx`.

The nondeterminism in the actual output (`#0` vs `#1..#4`) is exactly this race: the result depends on thread-block scheduling order.

### 3. Returned indices are not mapped back to the input

The host sorts the boxes by descending score first:

```cpp
auto order_t = std::get<1>(scores.sort(/*stable=*/true, 0, /*descending=*/true));
auto boxes_sorted = boxes.index_select(0, order_t).contiguous();
```

The kernel works on the sorted array, so `keep`'s indices are **positions in the sorted order**; but these are returned directly as the result without mapping through `order_t`. Whenever scores are not strictly decreasing, the output cannot be aligned with `torchvision.ops.nms`, which returns original indices.

## Suggested Fix

Split into two phases, decoupling "parallel IoU computation" from "sequential resolution":

1. **Kernel 1 (parallel, race-free)**: each thread computes the IoU for a `(i, j), j > i` pair and writes `iou[i][j] > tau` into a suppression bitmask via `atomicOr` (each thread only writes its own bit — no write conflicts);
2. **Kernel 2 or host side (sequential)**: resolve boxes in descending score order — if box i is not suppressed by any kept box, keep it and mark every box it suppresses in the bitmask. This phase is O(N) bitmask queries;
3. Map the keep results back to original indices via `order_t`;
4. Allocate the bitmask with `torch::zeros`.

### PR DESCRIPTION
**Pull request by chuoer47**
## Summary

Fixes #517.

`nms_kernel` in `kernels/nms/nms.cu` had three compounding correctness defects (details and minimal reproduction in the issue):

1. **Read of uninitialized memory**: `keep` was allocated with `torch::empty` but read before being written;
2. **Cross-block data race**: each thread scanned `keep[0..idx-1]` written by threads in other blocks with no synchronization - same input, different results across runs;
3. **Wrong index semantics**: returned indices referred to the score-sorted array and were never mapped back to the original input indices.

## Fix: two-phase kernel

- **Phase 1 (`nms_iou_mask_kernel`, parallel, race-free)**: one warp per box (boxes pre-sorted by descending score). The 32 lanes compute IoU for 32 consecutive candidates `j > i` in parallel, and `__ballot_sync` collects the verdicts into a single 32-bit word of a suppression bitmask. Each warp writes only its own bitmask row - no atomics, no write conflicts, and 32 IoU evaluations cost 1 global write. `j_base` is aligned to 32 so that bit k of word w is globally box `w*32 + k`.
- **Phase 2 (`nms_resolve_kernel`, sequential on one block)**: shared-memory bitmap, boxes resolved in descending score order; `__syncthreads` guarantees every decision sees all earlier ones - the happens-before that cross-block access can never provide. Kept boxes apply their suppression row with all threads in parallel (O(words) per round).
- Host side maps kept sorted positions back through the sort permutation, so the output matches `torchvision.ops.nms` (original indices). All buffers are `torch::zeros`-allocated.
- Complexity: O(N^2) IoU work (same as before, but now race-free); bitmask memory is N^2/8 bytes (documented; torchvision's tiled layout is the follow-up optimization, linked in README).

## Other changes

- `nms.py`: fixed reversed `(boxes, scores)` argument order in `run_benchmark` (the two implementations were being called with swapped arguments); added correctness checks that run before the benchmark - the fixed 6-box issue repro x5, and a random sweep (5 sizes x 3 seeds x 2 thresholds, with shuffled scores and forced ties) compared element-wise against `torchvision.ops.nms`.
- `nms.cc`: fixed compile errors (missing `;` after `struct Box`, undefined `tbox` -> `other`), added missing `<algorithm>` include, const lambda params.
- `README.md`: documents the two-phase design, the +1-pixel area convention of the CPU reference, and the new test output.

## Verification (RTX 4090, nvcc 12.4, torch 2.5.1+cu124, torchvision 0.20.1+cu124)

- Fixed 6-box case x5 runs: all match torchvision `[0, 2, 3, 4]`, deterministic (was nondeterministic before)
- Random sweep 5 sizes x 3 seeds x 2 thresholds: all element-wise identical to torchvision
- Edge cases pass: single box, no overlap, all-overlap, IoU exactly at threshold (`>` semantics)
- Benchmark (keep counts identical to torchvision at every size):

```
nboxes=1024  nms: 0.216ms   nms_th: 0.136ms
nboxes=2048  nms: 0.318ms   nms_th: 0.246ms
nboxes=4096  nms: 0.504ms   nms_th: 0.600ms
nboxes=8192  nms: 0.873ms   nms_th: 1.523ms  (1.7x faster)
```

### CODE REVIEW

File: kernels/nms/nms.cu [hunk 1]
```diff
@@ -67,33 +118,64 @@ __global__ void nms_kernel(const float *boxes, const float *scores, int *keep,
...
+  nms_resolve_kernel<<<1, resolve_threads, smem_bytes>>>(
+      reinterpret_cast<const unsigned int *>(mask.data_ptr()),
+      reinterpret_cast<int *>(keep.data_ptr()), num_boxes, mask_words);
+
+  // map kept sorted positions back to original input indices
+  auto keep_cpu = keep.to(torch::kCPU);
+  auto order_cpu = order_t.to(torch::kCPU);
   auto keep_accessor = keep_cpu.accessor<int, 1>();
+  auto order_accessor = order_cpu.accessor<long, 1>();
+  std::vector<int> keep_indices;
```
[the remark this answers is not in the document]
Comment (chuoer47) [reply kernels/nms/nms.cu]: Fixed in 951c09a: the binding now returns int64 indices on the same device as the inputs (matching torchvision.ops.nms), and accessor<long, 1> is replaced with data_ptr<int64_t>. check_correctness()'s torch.equal comparison now works directly without dtype/device conversion.

**Comment by chuoer47** [comment]
@DefTruth All three review rounds are addressed:

- Uninitialized keep + cross-block race + index semantics: fixed by the two-phase rewrite (25a96f7)
- Empty input (num_boxes == 0): handled (0658f27)
- Return dtype/device: now returns int64 indices on the same device as the inputs, matching torchvision.ops.nms; accessor<long, 1> replaced with data_ptr<int64_t> (951c09a). torch.equal against torchvision passes directly on all correctness checks (fixed 6-box repro x5, random sweep 5 sizes x 3 seeds x 2 thresholds).

Note: the nms.cc negative-IoU nit (clamp inner_h/inner_w) was left as-is intentionally since hard_nms only compares against the threshold - happy to add the clamp if you prefer.

Ready for another review whenever you have time. Thanks!

**Review by DefTruth — APPROVED** [review]
LGTM
